lib/spack/spack/package.py

```diff
@@ -1550,80 +1573,76 @@ def run_test(self, exe, options=[], expected=[], status=None,
+                    exe, runner, options, expected, status, installed, purpose)
+                print("PASSED")
+            except BaseException as e:
+                # print a summary of the error to the log file
+                # so that cdash and junit reporters know about it
+                exc_type, _, tb = sys.exc_info()
+                print('FAILED: %s' % e)
+                import traceback
+                # remove the current call frame to get back to
+                stack = traceback.extract_stack()[:-1]
+
+                # Package files have a line added at import time, so we re-read
+                # the file to make line numbers match. We have to subtract two
+                # from the line number because the original line number is
+                # inflated once by the import statement and the lines are
+                # displaced one by the import statement.
+                for i, entry in enumerate(stack):
+                    filename, lineno, function, text = entry
+                    if spack.paths.is_package_file(filename):
+                        with open(filename, 'r') as f:
+                            lines = f.readlines()
+                        text = lines[lineno - 2]
+                        stack[i] = (filename, lineno, function, text)
+
+                # Format the stack to print and print it
+                out = traceback.format_list(stack)
+                for line in out:
+                    print(line.rstrip('\n'))
+
+                if exc_type is spack.util.executable.ProcessError:
+                    out = StringIO()
+                    spack.build_environment.write_log_summary(
+                        out, 'test', self.test_log_file, last=1)
+                    m = out.getvalue()
+                else:
+                    # We're below the package context, so get context from
+                    # stack instead of from traceback.
+                    # The traceback is truncated here, so we can't use it to
+                    # traverse the stack.
+                    m = '\n'.join(spack.build_environment.get_package_context(
                         traceback.extract_stack()))
 
-            exc = e  # e is deleted after this block
+                exc = e  # e is deleted after this block
 
-            # If we fail fast, raise another error
-            if spack.config.get('config:fail_fast', False):
-                raise TestFailure([(exc, m)])
-            else:
-                self.test_failures.append((exc, m))
+                # If we fail fast, raise another error
+                if spack.config.get('config:fail_fast', False):
+                    raise TestFailure([(exc, m)])
+                else:
+                    self.test_failures.append((exc, m))
 
-    def _run_test_helper(self, exe, options, expected, status, installed,
-                         purpose):
+    def _run_test_helper(self, exe, runner, options, expected, status,
```

Comment:
We don't need to pass both `exe` and `runner`, we can always get the name of the executable from `runner.name`

Reply:
I hadn't looked that closely.  Will fix.  Thanks!

Reply:
Done.  Also changed the string commands from using '%s' to using `format` since this is new code and the latter is supposed to be the way forward.